vespalib/src/vespa/vespalib/datastore/compact_buffer_candidates.cpp

```diff
@@ -26,15 +29,23 @@ CompactBufferCandidates::add(uint32_t buffer_id, size_t used, size_t dead)
     _dead += dead;
 }
 
+void
+CompactBufferCandidates::set_free_buffers(uint32_t free_buffers)
+{
+    _free_buffers = free_buffers;
+}
+
 void
 CompactBufferCandidates::select(std::vector<uint32_t>& buffers)
 {
     if (_candidates.empty()) {
         return;
     }
-    if (_candidates.size() > _max_buffers) {
-        std::nth_element(_candidates.begin(), _candidates.begin() + (_max_buffers - 1), _candidates.end());
-        _candidates.resize(_max_buffers);
+    uint32_t max_buffers = ceil(std::min(_candidates.size() * _max_buffers_ratio, _free_buffers * 0.2));
```

Comment:
Consider adding a constexpr for the magic constant 0.2. Also consider adding a comment for lines 44-45 describing the overall idea for what we want to achieve here.